### PR TITLE
CSAM: Attach Self-Generated Report to Contact

### PR DESCRIPTION
**Pull request by GPaoloni**
Primary reviewer: @stephenhand 

## Description
This PR:
- Adds a migration to update `CSAMReports` table:
  - Adds column `reportType` (text) to specify the type of the report (`self-generated` or `counsellor-generated`).
  - Adds column `aknowledged` (boolean) to indicate if the report was aknowledged by the external organization. This is only relevant for `self-generated` reports, since we want to confirm that the record in the DB also has a match in the outer service. All `counsellor-generated` records are aknowledged at creation time, since they are saved in the DB **_after_** they are reported to IWF.
  - Update all existing records.
- Add endpoints and logic to support two new operations: `aknowledgeCsamReport` and `deleteCsamReport` (they do what the name says, nothing more :P).
- Extend support for `createCSAMReport` to accept `reportType` = `self-generated`. In this case we generate a UUID that will replace the `csamReportId` value (that is populated with the one generated by IWF in the case of `cousellor-generated` reports).

Q: Is desired that we filter out the reports if they are not acknowledged?

### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1407)
- [x] New tests added

### Verification steps
~A PR on the Flex side will be up tomorrow, which will make the testing way easier :)~
It's a draft, but you could use https://github.com/techmatters/hrm/pull/268 to see if this is working :) (see instructions there).
- Start containerized DB, run migrations, compile code and start local server.
- Confirm that the `POST` operation on `/csamReports` works as expected for the existing csam reports logic.
- Confirm that the `POST` operation on `/csamReports` works as expected when the payload contains `reportType` = `self-generated`.
- Confirm that the `PATCH` operation on `/csamReports/:reportId/aknowled` sets `aknowledged` = `true` (for `counsellor-generated` reports is a no-op).
- Confirm that the `DELETE` operation on `/csamReports/:reportId` deletes the contact if exists.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -13749,6 +13756,7 @@
         "@tech-matters/hrm-types": "^1.0.0",
         "@types/deep-diff": "^1.0.1",
         "@types/express": "^4.17.13",
+        "@types/http-errors": "*",
```
Comment (stephenhand) [inline package-lock.json:13759]: Is it weird that the version range generated here is `*` even though you set `^2.0.1` in the `package.josn`?

**Comment by GPaoloni** [comment]
@stephenhand I think it will be easier to do re-review this by commits.

File: package-lock.json (hunk 1 above)
Comment (GPaoloni) [reply package-lock.json:13841]: Should we sorry about this? Is just a dev dep for the types.

Comment (stephenhand) [reply package-lock.json:13841]: I'm just asking a Q, no need to change anything

File: hrm-service/src/csam-report/csam-report-data-access.ts [hunk 2]
```diff
@@ -1,54 +1,75 @@
...
+  acknowledged: boolean;
+  twilioWorkerId?: string;
+  csamReportId?: string;
+  contactId?: number;
 };
 
-export type CreateCSAMReport = Omit<NewCSAMReportRecord, 'accountSid'>;
+export type CSAMReport = NewCSAMReport & {
+  id: number;
+  accountSid: string;
```
Comment (stephenhand) [inline hrm-service/src/csam-report/csam-report-data-access.ts:21]: Do we need this as part of the type? We should only ever be dealing with reports from one account in HRM anywat

File: hrm-service/src/csam-report/csam-report-routes-v0.ts [hunk 3]
```diff
@@ -9,10 +10,35 @@ const csamReportRouter = SafeRouter();
...
+  }
+  if ((!reportType || reportType === "counsellor-generated") && !csamReportId) {
+    throw createError(422, 'Invalid, "reportType" argument specifies "counsellor-generated" report, but no csamReportId argument provided'); 
+  }
+
+  const createdCSAMReport = await createCSAMReport({ contactId, csamReportId, twilioWorkerId, reportType }, accountSid);
   res.json(createdCSAMReport);
 });
 
+csamReportRouter.post('/:reportId/acknowledge', publicEndpoint, async (req: Request & { accountSid: string }, res: Response) => {
```
Comment (stephenhand) [inline hrm-service/src/csam-report/csam-report-routes-v0.ts:27]:
```suggestion
csamReportRouter.post('/:reportId(\\d+)/acknowledge', publicEndpoint, async (req: Request & { accountSid: string }, res: Response) => {
```

We wouldn't have to validate the ID is a number in code then?

**Review by stephenhand — APPROVED** [review]
LGTM, couple of minor Qs, nothing blocking though

**Pull request merged**